src/kakarot/instructions/block_information.cairo

```diff
@@ -32,6 +36,64 @@ namespace BlockInformation {
     const GAS_COST_SELFBALANCE = 5;
     const GAS_COST_BASEFEE = 2;
 
+    // @notice COINBASE operation.
+    // @dev Get the hash of one of the 256 most recent complete blocks.
+    // @dev 0 if the block number is not in the valid range.
+    // @custom:since Frontier
+    // @custom:group Block Information
+    // @custom:gas 20
+    // @custom:stack_consumed_elements 1
+    // @custom:stack_produced_elements 1
+    // @param ctx The pointer to the execution context
+    // @return The pointer to the updated execution context.
+    func exec_blockhash{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
```

Comment:
This is the BLOCKHASH opcode implementation. It creates a `block_context_dict` dictionary from the array of block numbers and hashes from `BlockContext` and uses `dict_read` to lookup the blockhash value sing the stack input (block number) as key. It returns 0 if the block number is not within the valid range